services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleAssociationScannerJobManagerImpl.java

```diff
@@ -180,17 +193,40 @@ private static int getMessageDelay() {
 	}
 	
 	private int flushRecordsBatch(Set<FileHandleAssociationRecord> recordsBatch) throws RecoverableMessageException {
-		validateStackReadWrite();
+		if (recordsBatch.isEmpty()) {
+			return 0;
+		}
+		
 		int size = recordsBatch.size();
+		
 		try {
 			kinesisLogger.logBatch(FileHandleAssociationRecord.STREAM_NAME, new ArrayList<>(recordsBatch));
 		} catch (AwsKinesisDeliveryException e) {
 			throw new RecoverableMessageException(e);
 		}
-		recordsBatch.clear();
+		
 		return size;
 	}
 	
+	private int relinkRecords(Set<FileHandleAssociationRecord> recordsBatch) {
+		
+		if (recordsBatch.isEmpty()) {
+			return 0;
+		}
+		
+		Set<Long> fileHandleIds = recordsBatch.stream()
+				.map(FileHandleAssociationRecord::getFileHandleId)
+				.collect(Collectors.toSet());
+		
+		List<Long> updatedIds = fileHandleDao.updateBatchStatus(new ArrayList<>(fileHandleIds), FileHandleStatus.AVAILABLE, FileHandleStatus.UNLINKED, 0);
```

Comment:
try a  query to find one row that would need updated from the batch, then do the update, we minimize write locks.  Also it should be super rare to have a hit.